Sporadic test failure: TestChannelAccessChanges (integration test mode)

In http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/107/console:

```
2017-11-14T11:40:53.357-08:00 Changes+: MultiChangesFeed sending {Seq:4, ID:gamma, Changes:[map[rev:1-4ceecff40b4726ca5e5ecddc2d988963]]}   (to alice)
2017-11-14T11:40:53.357-08:00 Changes+: Channel feed processing seq:5 in channel beta   (to alice)
2017-11-14T11:40:53.357-08:00 Changes+: Channel feed processing seq:6 in channel beta   (to alice)
2017-11-14T11:40:53.357-08:00 Changes+: Channel feed processing seq:7 in channel beta   (to alice)
2017-11-14T11:40:53.357-08:00 Changes+: MultiChangesFeed sending {Seq:5, ID:a1, Changes:[map[rev:1-f2c898cf01dd5f572ac9b28d58ba7d4b]]}   (to alice)
2017-11-14T11:40:53.357-08:00 Changes+: MultiChangesFeed sending {Seq:6, ID:b1, Changes:[map[rev:1-7dcf5f78478f92454f1d0b359f204cbc]]}   (to alice)
2017-11-14T11:40:53.357-08:00 Changes+: MultiChangesFeed sending {Seq:7, ID:d1, Changes:[map[rev:1-d1ece69976d8cb66ff2a77210094db7c]]}   (to alice)
2017-11-14T11:40:53.357-08:00 Changes+: Channel feed processing seq:8 in channel beta   (to alice)
2017-11-14T11:40:53.357-08:00 Changes+: Channel feed processing seq:9 in channel beta   (to alice)
2017-11-14T11:40:53.357-08:00 Changes+: Channel feed processing seq:10 in channel beta   (to alice)
2017-11-14T11:40:53.357-08:00 Changes+: MultiChangesFeed sending {Seq:8, ID:g1, Changes:[map[rev:1-97023f8d0f6031c1587c483ff7f5a2ac]]}   (to alice)
2017-11-14T11:40:53.357-08:00 Changes+: MultiChangesFeed sending {Seq:9, ID:alpha, Changes:[map[rev:2-c3bcc88e1ec141c7c74f9b73d8363007]]}   (to alice)
2017-11-14T11:40:53.357-08:00 Changes+: Found sequence later than stable sequence: stable:[9] entry:[10] (epsilon)
2017-11-14T11:40:53.357-08:00 Changes: MultiChangesFeed done   (to alice)
2017-11-14 11:40:53.357770 I | _changes looks like: {"results":[
{"seq":2,"id":"beta","changes":[{"rev":"1-ffe7f169bea54c435a7ca90bccf7ab95"}]}
,{"seq":3,"id":"delta","changes":[{"rev":"1-742d9f0eb1f4fa613be657819a5df5a2"}]}
,{"seq":4,"id":"gamma","changes":[{"rev":"1-4ceecff40b4726ca5e5ecddc2d988963"}]}
,{"seq":5,"id":"a1","changes":[{"rev":"1-f2c898cf01dd5f572ac9b28d58ba7d4b"}]}
,{"seq":6,"id":"b1","changes":[{"rev":"1-7dcf5f78478f92454f1d0b359f204cbc"}]}
,{"seq":7,"id":"d1","changes":[{"rev":"1-d1ece69976d8cb66ff2a77210094db7c"}]}
,{"seq":8,"id":"g1","changes":[{"rev":"1-97023f8d0f6031c1587c483ff7f5a2ac"}]}
,{"seq":9,"id":"alpha","changes":[{"rev":"2-c3bcc88e1ec141c7c74f9b73d8363007"}]}
],
"last_seq":"9"}

	api_test.go:1805
	assert.Equals(t, len(changes.Results), len(expectedIDs))	

	expected: 9
	     got: 8
2017-11-14T11:40:53.358-08:00 Changes+: changeListener.Stop() called
--- FAIL: TestChannelAccessChanges (9.88s)
	assert.go:31: 
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 9584 [running]:
testing.tRunner.func1(0xc4216ce3c0)
	/root/.gvm/gos/go1.9.2/src/testing/testing.go:711 +0x2d2
panic(0xbcbea0, 0x12aa6f0)
	/root/.gvm/gos/go1.9.2/src/runtime/panic.go:491 +0x283
github.com/couchbase/sync_gateway/rest.TestChannelAccessChanges(0xc4216ce3c0)
	/var/jenkins/workspace/sync-gateway-integration-master/godeps/src/github.com/couchbase/sync_gateway/rest/api_test.go:1809 +0x2ad9
testing.tRunner(0xc4216ce3c0, 0xcea0f0)
	/root/.gvm/gos/go1.9.2/src/testing/testing.go:746 +0xd0
created by testing.(*T).Run
	/root/.gvm/gos/go1.9.2/src/testing/testing.go:789 +0x2de
exit status 2
FAIL	github.com/couchbase/sync_gateway/rest	336.879s
```

--- a/rest/api_test.go
+++ b/rest/api_test.go
@@ -1734,10 +1734,10 @@ func TestChannelAccessChanges(t *testing.T) {
 		t,
 		alice.Channels(),
 		channels.TimedSet{
-			"!":     channels.NewVbSimpleSequence(0x1),
-			"zero":  channels.NewVbSimpleSequence(0x1),
-			"alpha": channels.NewVbSimpleSequence(0x1),
-			"delta": channels.NewVbSimpleSequence(0x3),
+			"!":     channels.NewVbSimpleSequence(uint64(1)),
+			"zero":  channels.NewVbSimpleSequence(uint64(1)),
+			"alpha": channels.NewVbSimpleSequence(uint64(1)),
+			"delta": channels.NewVbSimpleSequence(uint64(3)),
 		})
 
 	zegpold, _ = a.GetUser("zegpold")
@@ -1745,9 +1745,9 @@ func TestChannelAccessChanges(t *testing.T) {
 		t,
 		zegpold.Channels(),
 		channels.TimedSet{
-			"!":     channels.NewVbSimpleSequence(0x1),
-			"zero":  channels.NewVbSimpleSequence(0x1),
-			"gamma": channels.NewVbSimpleSequence(0x4),
+			"!":     channels.NewVbSimpleSequence(uint64(1)),
+			"zero":  channels.NewVbSimpleSequence(uint64(1)),
+			"gamma": channels.NewVbSimpleSequence(uint64(4)),
 		})
 
 	// Update a document to revoke access to alice and grant it to zegpold:
@@ -1760,9 +1760,9 @@ func TestChannelAccessChanges(t *testing.T) {
 		t,
 		alice.Channels(),
 		channels.TimedSet{
-			"!":     channels.NewVbSimpleSequence(0x1),
-			"zero":  channels.NewVbSimpleSequence(0x1),
-			"delta": channels.NewVbSimpleSequence(0x3),
+			"!":     channels.NewVbSimpleSequence(uint64(1)),
+			"zero":  channels.NewVbSimpleSequence(uint64(1)),
+			"delta": channels.NewVbSimpleSequence(uint64(3)),
 		})
 
 	zegpold, _ = a.GetUser("zegpold")
@@ -1770,10 +1770,10 @@ func TestChannelAccessChanges(t *testing.T) {
 		t,
 		zegpold.Channels(),
 		channels.TimedSet{
-			"!":     channels.NewVbSimpleSequence(0x1),
-			"zero":  channels.NewVbSimpleSequence(0x1),
-			"alpha": channels.NewVbSimpleSequence(0x9),
-			"gamma": channels.NewVbSimpleSequence(0x4),
+			"!":     channels.NewVbSimpleSequence(uint64(1)),
+			"zero":  channels.NewVbSimpleSequence(uint64(1)),
+			"alpha": channels.NewVbSimpleSequence(uint64(9)),
+			"gamma": channels.NewVbSimpleSequence(uint64(4)),
 		})
 
 	rt.MustWaitForDoc("alpha", t)
@@ -1809,29 +1809,18 @@ func TestChannelAccessChanges(t *testing.T) {
 	assert.Equals(t, changes.Results[0].ID, "a1")
 
 	// What happens if we call access() with a nonexistent username?
-	assertStatus(t, rt.Send(request("PUT", "/db/epsilon", `{"owner":"waldo"}`)), 201)
+	assertStatus(t, rt.Send(request("PUT", "/db/epsilon", `{"owner":"waldo"}`)), 201)  // seq 10
 
-	// Wait for change caching to complete before running resync below
-	// time.Sleep(500 * time.Millisecond)
+	// Must wait for sequence to arrive in cache, since the cache processor will be paused when UpdateSyncFun() is called
+	// below, which could lead to a data race if the cache processor is paused while it's processing a change
+	rt.ServerContext().Database("db").WaitForSequence(uint64(10))
 
 	// Finally, throw a wrench in the works by changing the sync fn. Note that normally this wouldn't
 	// be changed while the database is in use (only when it's re-opened) but for testing purposes
 	// we do it now because we can't close and re-open an ephemeral Walrus database.
 	dbc := rt.ServerContext().Database("db")
 	database, _ := db.GetDatabase(dbc, nil)
-
-	time.Sleep(time.Second)
-
-	validFrom, cachedEntries := database.GetChangeIndex().GetCachedChanges(
-		"alpha",
-		db.ChangesOptions{
-			Since: db.SequenceID{Seq: 0},
-		})
-
-
-	log.Printf("GetChanges(alpha) validFrom: %v.  cachedEntries: %v", validFrom, cachedEntries)
-
-
+	
 	changed, err := database.UpdateSyncFun(`function(doc) {access("alice", "beta");channel("beta");}`)
 	assert.Equals(t, err, nil)
 	assert.True(t, changed)

--- a/rest/api_test.go
+++ b/rest/api_test.go
@@ -1822,13 +1822,15 @@ func TestChannelAccessChanges(t *testing.T) {
 
 	time.Sleep(time.Second)
 
-	entries, err := database.GetChangeIndex().GetChanges(
-		"ABC",
+	validFrom, cachedEntries := database.GetChangeIndex().GetCachedChanges(
+		"alpha",
 		db.ChangesOptions{
 			Since: db.SequenceID{Seq: 0},
 		})
-	assertNoError(t, err, "Unexpected error")
-	log.Printf("GetChanges() entries: %v", entries)
+
+
+	log.Printf("GetChanges(alpha) validFrom: %v.  cachedEntries: %v", validFrom, cachedEntries)
+
 
 	changed, err := database.UpdateSyncFun(`function(doc) {access("alice", "beta");channel("beta");}`)
 	assert.Equals(t, err, nil)

--- a/rest/api_test.go
+++ b/rest/api_test.go
@@ -962,7 +962,7 @@ func TestBulkDocsUnusedSequencesMultiRevDoc(t *testing.T) {
 			Username: username,
 			Password: password,
 		},
-		Name: "db",
+		Name:         "db",
 		EnableXattrs: &rt1UseXattrs,
 	})
 
@@ -1066,7 +1066,7 @@ func TestBulkDocsUnusedSequencesMultiRevDoc2SG(t *testing.T) {
 			Username: username,
 			Password: password,
 		},
-		Name: "db",
+		Name:         "db",
 		EnableXattrs: &rt1UseXattrs,
 	})
 
@@ -1680,6 +1680,7 @@ func testAccessControl(t *testing.T, rt indexTester) {
 }
 
 func TestChannelAccessChanges(t *testing.T) {
+
 	base.ParseLogFlags([]string{"Cache", "Changes+", "CRUD", "DIndex+"})
 
 	rt := RestTester{SyncFn: `function(doc) {access(doc.owner, doc._id);channel(doc.channel)}`}
@@ -1729,9 +1730,25 @@ func TestChannelAccessChanges(t *testing.T) {
 
 	// Check user access:
 	alice, _ = a.GetUser("alice")
-	assert.DeepEquals(t, alice.Channels(), channels.TimedSet{"!": channels.NewVbSimpleSequence(0x1), "zero": channels.NewVbSimpleSequence(0x1), "alpha": channels.NewVbSimpleSequence(0x1), "delta": channels.NewVbSimpleSequence(0x3)})
+	assert.DeepEquals(
+		t,
+		alice.Channels(),
+		channels.TimedSet{
+			"!":     channels.NewVbSimpleSequence(0x1),
+			"zero":  channels.NewVbSimpleSequence(0x1),
+			"alpha": channels.NewVbSimpleSequence(0x1),
+			"delta": channels.NewVbSimpleSequence(0x3),
+		})
+
 	zegpold, _ = a.GetUser("zegpold")
-	assert.DeepEquals(t, zegpold.Channels(), channels.TimedSet{"!": channels.NewVbSimpleSequence(0x1), "zero": channels.NewVbSimpleSequence(0x1), "gamma": channels.NewVbSimpleSequence(0x4)})
+	assert.DeepEquals(
+		t,
+		zegpold.Channels(),
+		channels.TimedSet{
+			"!":     channels.NewVbSimpleSequence(0x1),
+			"zero":  channels.NewVbSimpleSequence(0x1),
+			"gamma": channels.NewVbSimpleSequence(0x4),
+		})
 
 	// Update a document to revoke access to alice and grant it to zegpold:
 	str := fmt.Sprintf(`{"owner":"zegpold", "_rev":%q}`, alphaRevID)
@@ -1739,9 +1756,25 @@ func TestChannelAccessChanges(t *testing.T) {
 
 	// Check user access again:
 	alice, _ = a.GetUser("alice")
-	assert.DeepEquals(t, alice.Channels(), channels.TimedSet{"!": channels.NewVbSimpleSequence(0x1), "zero": channels.NewVbSimpleSequence(0x1), "delta": channels.NewVbSimpleSequence(0x3)})
+	assert.DeepEquals(
+		t,
+		alice.Channels(),
+		channels.TimedSet{
+			"!":     channels.NewVbSimpleSequence(0x1),
+			"zero":  channels.NewVbSimpleSequence(0x1),
+			"delta": channels.NewVbSimpleSequence(0x3),
+		})
+
 	zegpold, _ = a.GetUser("zegpold")
-	assert.DeepEquals(t, zegpold.Channels(), channels.TimedSet{"!": channels.NewVbSimpleSequence(0x1), "zero": channels.NewVbSimpleSequence(0x1), "alpha": channels.NewVbSimpleSequence(0x9), "gamma": channels.NewVbSimpleSequence(0x4)})
+	assert.DeepEquals(
+		t,
+		zegpold.Channels(),
+		channels.TimedSet{
+			"!":     channels.NewVbSimpleSequence(0x1),
+			"zero":  channels.NewVbSimpleSequence(0x1),
+			"alpha": channels.NewVbSimpleSequence(0x9),
+			"gamma": channels.NewVbSimpleSequence(0x4),
+		})
 
 	rt.MustWaitForDoc("alpha", t)
 
@@ -1779,42 +1812,50 @@ func TestChannelAccessChanges(t *testing.T) {
 	assertStatus(t, rt.Send(request("PUT", "/db/epsilon", `{"owner":"waldo"}`)), 201)
 
 	// Wait for change caching to complete before running resync below
-	time.Sleep(500 * time.Millisecond)
+	// time.Sleep(500 * time.Millisecond)
 
 	// Finally, throw a wrench in the works by changing the sync fn. Note that normally this wouldn't
 	// be changed while the database is in use (only when it's re-opened) but for testing purposes
 	// we do it now because we can't close and re-open an ephemeral Walrus database.
 	dbc := rt.ServerContext().Database("db")
-	db, _ := db.GetDatabase(dbc, nil)
-	changed, err := db.UpdateSyncFun(`function(doc) {access("alice", "beta");channel("beta");}`)
+	database, _ := db.GetDatabase(dbc, nil)
+
+	time.Sleep(time.Second)
+
+	entries, err := database.GetChangeIndex().GetChanges(
+		"ABC",
+		db.ChangesOptions{
+			Since: db.SequenceID{Seq: 0},
+		})
+	assertNoError(t, err, "Unexpected error")
+	log.Printf("GetChanges() entries: %v", entries)
+
+	changed, err := database.UpdateSyncFun(`function(doc) {access("alice", "beta");channel("beta");}`)
 	assert.Equals(t, err, nil)
 	assert.True(t, changed)
-	changeCount, err := db.UpdateAllDocChannels(true, false)
+	changeCount, err := database.UpdateAllDocChannels(true, false)
 	assert.Equals(t, err, nil)
 	assert.Equals(t, changeCount, 9)
 
-	// Artificial delay to let updates show up on changes feed
-	// TODO: not sure how to change this to wait for a sequence
-	time.Sleep(time.Second)
-
-	changes = changesResults{}
 	expectedIDs := []string{"beta", "delta", "gamma", "a1", "b1", "d1", "g1", "alpha", "epsilon"}
-	response = rt.Send(requestByUser("GET", "/db/_changes", "", "alice"))
-	json.Unmarshal(response.Body.Bytes(), &changes)
-	log.Printf("_changes looks like: %s", response.Body.Bytes())
+	changes, err = rt.WaitForChanges(len(expectedIDs), "/db/_changes", "alice")
+	assertNoError(t, err, "Unexpected error")
+	log.Printf("_changes looks like: %+v", changes)
 	assert.Equals(t, len(changes.Results), len(expectedIDs))
-	json.Unmarshal(response.Body.Bytes(), &changes)
 
 	for i, expectedID := range expectedIDs {
+		if changes.Results[i].ID != expectedID {
+			log.Printf("changes.Results[i].ID != expectedID.  changes.Results: %+v, expectedIDs: %v", changes.Results, expectedIDs)
+		}
 		assert.Equals(t, changes.Results[i].ID, expectedID)
 	}
 
 	// Check accumulated statistics:
-	assert.Equals(t, db.ChangesClientStats.TotalCount(), uint32(5))
-	assert.Equals(t, db.ChangesClientStats.MaxCount(), uint32(1))
-	db.ChangesClientStats.Reset()
-	assert.Equals(t, db.ChangesClientStats.TotalCount(), uint32(0))
-	assert.Equals(t, db.ChangesClientStats.MaxCount(), uint32(0))
+	assert.Equals(t, database.ChangesClientStats.TotalCount(), uint32(5))
+	assert.Equals(t, database.ChangesClientStats.MaxCount(), uint32(1))
+	database.ChangesClientStats.Reset()
+	assert.Equals(t, database.ChangesClientStats.TotalCount(), uint32(0))
+	assert.Equals(t, database.ChangesClientStats.MaxCount(), uint32(0))
 }
 
 func TestAccessOnTombstone(t *testing.T) {

--- a/rest/api_test.go
+++ b/rest/api_test.go
@@ -1809,7 +1809,7 @@ func TestChannelAccessChanges(t *testing.T) {
 	assert.Equals(t, changes.Results[0].ID, "a1")
 
 	// What happens if we call access() with a nonexistent username?
-	assertStatus(t, rt.Send(request("PUT", "/db/epsilon", `{"owner":"waldo"}`)), 201)  // seq 10
+	assertStatus(t, rt.Send(request("PUT", "/db/epsilon", `{"owner":"waldo"}`)), 201) // seq 10
 
 	// Must wait for sequence to arrive in cache, since the cache processor will be paused when UpdateSyncFun() is called
 	// below, which could lead to a data race if the cache processor is paused while it's processing a change
@@ -1820,7 +1820,7 @@ func TestChannelAccessChanges(t *testing.T) {
 	// we do it now because we can't close and re-open an ephemeral Walrus database.
 	dbc := rt.ServerContext().Database("db")
 	database, _ := db.GetDatabase(dbc, nil)
-	
+
 	changed, err := database.UpdateSyncFun(`function(doc) {access("alice", "beta");channel("beta");}`)
 	assert.Equals(t, err, nil)
 	assert.True(t, changed)

